### PR TITLE
fix issue with missing metadata signature on custom repositories

### PR DESCRIPTION
**Pull request by Shazootua9sh**
The existing code breaks mirrors of custom repositories, which don't provide the gpg public signature key at the expected location ( repodata/repomd.xml.key ) .
If the key is not found, the code stops downloading the gpg signature of the metadata ( repodata/repomd.xml.asc ) . Without the metadata signature, the mirrored repository is useless.

Proposed fix: 
- RMT should download the metadata signature, nevermind if the corresponding gpg public key is present. It should not be the task of a mirroring tool to verify the signature; this is the task of the package management tools (e.g. zypper).
- the proposed fix first downloads the signature, afterwards the key. if the key is not found, a warning gets issued, but the mirroring process is not interruped and the mirror is kept in  a wporking state.

Examples of repositories which are affected:
- https://packages.icinga.com/SUSE/15.0/release/
- https://packages.icinga.com/SUSE/sles12-sp4/release/
- https://repo.mongodb.org/zypper/suse/12/mongodb-org/4.0/x86_64/

### CODE REVIEW

**Comment by Shazootua9sh** [comment]
Travis CI test:
"RMT::Mirror when GPG signature is incomplete when one of the signature files is missing raises RMT::Mirror::Exception"
In my opinion, this is a wrong assumption: a signature is not incomplete just because the signing key is distributed by other channels.

**Comment by ikapelyukhin** [comment]
Hi, @Shazootua9sh.

The order of downloading `repomd.xml.key` and `repomd.xml.asc` doesn't really matter because of the following code:

https://github.com/SUSE/rmt/blob/56e51957b5a1691c756f1ae44b122f6ba20c5532/lib/rmt/mirror.rb#L103

Currently we either expect either both `key` and `asc` files to be present, or none at all.

Both `key` and `asc` files are present in all openSUSE and SLES repos. We've had issues in the past when signatures wouldn't be downloaded correctly from the CDN and that would cause issues on the client machines.

It would be really nice if 3rd party software vendors would conform to openSUSE packaging standards and publish the `key` file :slightly_smiling_face:

**Comment by Shazootua9sh** [comment]
Hi @ikapelyukhin ,

> The order of downloading `repomd.xml.key` and `repomd.xml.asc` doesn't really matter because of the following code:

The order of the downloading matters a lot: what happens with the existing code, is:
- RMT tries to download the `key ` ; If there is none, the code throws an exception with return code 404. Because of the exception,  RMT doesn't even try to download the `asc`.

https://github.com/SUSE/rmt/blob/56e51957b5a1691c756f1ae44b122f6ba20c5532/lib/rmt/mirror.rb#L95-L97

If we change the order of the download process:
- first the `asc` is downloaded; works fine
- then the `key` is downloaded, and throws the exception.

at this point we already have the `asc` . The mirrored repo has everything it needs.

Afterwards the code you cited comes in:
> https://github.com/SUSE/rmt/blob/56e51957b5a1691c756f1ae44b122f6ba20c5532/lib/rmt/mirror.rb#L103

> Currently we either expect either both `key` and `asc` files to be present, or none at all.

that's not entirely true: you expect that:
- if there is a `key`, there has to be a `asc`
in our case, we only have an `asc`, but no key, so there is no problem, at least not from the code.

> Both `key` and `asc` files are present in all openSUSE and SLES repos. We've had issues in the past when signatures wouldn't be downloaded correctly from the CDN and that would cause issues on the client machines.
> 
> It would be really nice if 3rd party software vendors would conform to openSUSE packaging standards and publish the `key` file slightly_smiling_face

As far as I know, the specification even for SUSE repos doesn't require, that the signing key is present in the repository:
[https://en.opensuse.org/openSUSE:Standards_Rpm_Metadata](https://en.opensuse.org/openSUSE:Standards_Rpm_Metadata)

> Based on this model, the master index is repomd.xml, and the signature has to be provided as repomd.xml.asc.
>
> The public part of the key used to sign the master index **can be provided** as repomd.xml.key 

I run the code as i proposed in production, because we need to mirror some custom repos. The original code renders all of our custom repos useless, because of the missing metadata signature.
The mirroring of the SLES repos is not affected negativly, as ar as aÍ can tell.

**Comment by ikapelyukhin** [comment]
> As far as I know, the specification even for SUSE repos doesn't require, that the signing key is present in the repository:
> https://en.opensuse.org/openSUSE:Standards_Rpm_Metadata#Metadata_Signature_and_Checksum
> 
> > Based on this model, the master index is repomd.xml, and the signature has to be provided as repomd.xml.asc.
> > The public part of the key used to sign the master index **can be provided** as repomd.xml.key

OK, fair enough.

> The order of downloading matters a lot: what happens with the existing code, is:
> https://github.com/SUSE/rmt/blob/56e51957b5a1691c756f1ae44b122f6ba20c5532/lib/rmt/mirror.rb#L95-L97

OK, this code actually wouldn't be needed at all, because the condition would never be triggered.
Please drop this exception and the respective spec (the one currently red in Travis):

https://github.com/SUSE/rmt/blob/8223a4beb0d62ed620f2cd4543e26a63a41c5cf8/spec/lib/rmt/mirror_spec.rb#L575-L596

> The mirroring of the SLES repos is not affected negativly, as ar as I can tell.

Hmm, I'm not sure about that. But weird things could have happened in the current implementation, different kind of weird things could happen with your proposed changes, but I suppose the likelihood of those things happening is about the same.

But perhaps if we see GPG issues popping back, we could try to solve them by other means (e.g., by adding some option like `require_gpg_signatures` for RMT or maybe try and solve it on SCC side).

File: spec/lib/rmt/mirror_spec.rb [hunk 1]
```diff
@@ -554,7 +554,7 @@
       FileUtils.remove_entry(@tmp_dir)
     end
 
-    context 'when signatures do not exist' do
+    context 'when the signing key does not exist' do
```
Comment (Shazootua9sh) [inline spec/lib/rmt/mirror_spec.rb]: clarify what is the problem

File: lib/rmt/mirror.rb [hunk 2]
```diff
@@ -86,21 +86,17 @@ def mirror_metadata
...
-      if (e.http_code == 404)
-        @logger.info(_('Repository metadata signatures are missing'))
-      else
-        raise(_('Failed to get repository metadata signatures with HTTP code %{http_code}') % { http_code: e.http_code })
-      end
+      @logger.info(_('Repository metadata signatures are missing'))
     end
 
-    raise _('The repository has an incomplete metadata signature') if (key_file && !signature_file)
+    raise _('The repository has a missing metadata signature') if (!signature_file)
```
[the remark this answers is not in the document]
Comment (Shazootua9sh) [reply lib/rmt/mirror.rb]: ok, so completly drop the exception?

[the remark this answers is not in the document]
Comment (ikapelyukhin) [reply lib/rmt/mirror.rb]: Yep.

**Comment by Shazootua9sh** [comment]
hm, I don't get why Travis fails now. I already reverted most of the changes I inctroduced.

**Comment by ikapelyukhin** [comment]
@Shazootua9sh delete the whole `context` block: https://github.com/SUSE/rmt/blob/8223a4beb0d62ed620f2cd4543e26a63a41c5cf8/spec/lib/rmt/mirror_spec.rb#L575-L596

(There are some `expect` statements there, but you've removed the call to `mirror`, so they fail)

**Comment by Shazootua9sh** [comment]
> (There are some expect statements there, but you've removed the call to rmt_mirror.mirror, so they fail) 

ay, thanks.
So now just the exception and the corresponding specs are removed.

I'd like not to touch the error handling of the downloader, because it's used in other specs:
https://github.com/SUSE/rmt/blob/56e51957b5a1691c756f1ae44b122f6ba20c5532/lib/rmt/mirror.rb#L95-L99

**Comment by ikapelyukhin** [comment]
> So now just the exception and the corresponding specs are removed.

Yep, CI is green now :+1: 

> I'd like not to touch the error handling of the downloader, because it's used in other specs:

Yep, that should be fine.

**Comment by Shazootua9sh** [comment]
nice, thanks for your help and guidance.
should I do anything eles, like tidy up my commit history?

**Comment by ikapelyukhin** [comment]
> nice, thanks for your help and guidance.
> should I do anything eles, like tidy up my commit history?

No problem. Sure, feel free to squash some commits (alternatively we could do squash on merge).

**Comment by ikapelyukhin** [comment]
@fschueller it probably would be a good idea to reflect this change in the changes file.

**Pull request merged**